add Tree-Path KL Divergence loss for hier classification + unit test

Jyc323 opened the pull request:
### Summary
Introduced a new hierarchical regularization loss `TreePathKLDivergenceLoss` that enforces consistency across hierarchical label predictions by minimizing KL divergence between model probabilities and a constructed target distribution along the true label path.

### Key Changes
- Added `tree_path_kl_divergence.py` under `lib/src/otx/backend/native/models/common/losses/`
  - Implements `TreePathKLDivergenceLoss` with configurable reduction (`batchmean`, `mean`, `sum`) and weighting.
  - Target distribution assigns equal probability (1/L) to the GT class at each hierarchy level.
- Added comprehensive unit tests in `lib/tests/unit/backend/native/models/common/losses/test_tree_path_kl_divergence.py` covering:
  - Shape and finiteness checks across multiple levels.
  - Gradient flow validation to ensure backprop works correctly.
  - Alignment vs misalignment sanity check: aligned logits give lower loss.
  - **Exact-value correctness** for:
    - Single-level case (loss equals `CrossEntropyLoss`).
    - Multi-level case (manual KL computation matches PyTorch loss).

### Motivation
Existing metrics capture evaluation aspects of hierarchical classification, but training lacked a regularizer that explicitly
encourages consistency across levels. The Tree-Path KL loss:
- Penalizes predictions that scatter probability mass off the ground-truth path.
- Provides a smooth differentiable signal to improve hierarchical consistency beyond hard CE losses.

This makes the training loop more robust for hierarchical models, reducing path-inconsistent predictions and improving downstream metrics.

### Preliminary Results
We evaluated the effect of Tree-Path KL loss on multiple hierarchical datasets using **EfficientNet_v2 (timm model)** from OTX. The KL weight (λ) was varied between 0 (baseline), 1, and 5. Results show that incorporating KL improves performance on most datasets.

| Dataset          | Metric             | λ=0  | λ=1  | λ=5  |
|------------------|--------------------|-------|-------|-------|
| **CUB_200_2011** | Accuracy           | 0.853 | 0.850 | **0.858** |
|                  | Full Path Accuracy | 0.735 | 0.724 | **0.764** |
|                  | Leaf Accuracy      | **0.825** | 0.821 | 0.812 |
|                  | Weighted Precision | **0.836** | 0.832 | 0.829 |
| **AmazonParrot** | Accuracy           | 0.718 | 0.725 | **0.742** |
|                  | Full Path Accuracy | 0.611 | 0.614 | **0.659** |
|                  | Leaf Accuracy      | 0.333 | 0.334 | **0.346** |
|                  | Weighted Precision | 0.490 | **0.504** | **0.504** |
| **FGVC-Aircraft**| Accuracy           | 0.709 | 0.742 | **0.764** |
|                  | Full Path Accuracy | 0.551 | 0.593 | **0.629** |
|                  | Leaf Accuracy      | 0.614 | 0.639 | **0.660** |
|                  | Weighted Precision | 0.705 | 0.730 | **0.746** |
| **Fish-Vista**   | Accuracy           | 0.460 | 0.497 | **0.507** |
|                  | Full Path Accuracy | 0.217 | 0.264 | **0.280** |
|                  | Leaf Accuracy      | 0.416 | 0.452 | **0.458** |
|                  | Weighted Precision | 0.399 | 0.435 | **0.438** |

**Observations:**
- Increasing KL weight to 5 often improves **Full Path Accuracy**, showing stronger hierarchical consistency.
- Gains are consistent across datasets of varying granularity (birds, parrots, aircraft, fish).
- Some trade-off exists in **Leaf Accuracy** (CUB shows slight drop), but overall performance trends upward in hierarchy-sensitive metrics.

These results indicate that Tree-Path KL is a promising regularizer for hierarchical classification tasks.

### How to test

<!-- Describe the testing procedure for reviewers, if changes are
not fully covered by unit tests or manual testing can be complicated. -->
```
 pytest -q lib/tests/unit/backend/native/models/classification/losses/test_tree_path_kl_divergence.py
```

### Checklist

<!-- Put an 'x' in all the boxes that apply -->

- [x] I have added unit tests to cover my changes.​
- [x] I have added integration tests to cover my changes.​
- [x] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/open-edge-platform/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/open-edge-platform/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [x] I submit _my code changes_ under the same [Apache License](https://github.com/open-edge-platform/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2025 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

Jyc323 committed:
add Tree-Path KL Divergence loss for hier classification + unit test

rajeshgangireddy commented:
PS : This and  #4689  are PRs from our GSOC contributor.
Part of the project was 'H-label classification for VLMs' but the metrics and loss functions are useful and would be a nice addition to OTX.

sovrasov commented on lib/tests/unit/backend/native/models/classification/losses/test_tree_path_kl_divergence.py:
```diff
@@ -0,0 +1,134 @@
+
+    assert torch.allclose(test_kl.double(), manual_kl, atol=1e-8), (
+        f"manual={manual_kl.item():.12f} vs loss={test_kl.item():.12f}"
+    )
```
Missing new line at EOF

sovrasov commented on lib/src/otx/backend/native/models/classification/losses/tree_path_KL_divergence_loss.py:
```diff
@@ -0,0 +1,53 @@
+"""Module for defining TreePathKLDivergenceLoss."""
+
+from __future__ import annotations
+from typing import List
```
We're using python > 3.10, there is no need to import `List`: built-in `list` can be used instead

sovrasov commented on lib/src/otx/backend/native/models/classification/losses/tree_path_KL_divergence_loss.py:
```diff
@@ -0,0 +1,53 @@
+# Copyright (C) 2022 Intel Corporation
```
2025

sovrasov commented on lib/tests/unit/backend/native/models/classification/losses/test_tree_path_kl_divergence.py:
```diff
@@ -0,0 +1,134 @@
+import torch
```
License header is missing

sovrasov commented on lib/src/otx/backend/native/models/classification/losses/tree_path_KL_divergence_loss.py:
```diff
@@ -0,0 +1,53 @@
+    and 0 elsewhere, then uses KLDivLoss(log_softmax(logits), target_probs).
+    """
+
+    def __init__(self, reduction: str = "batchmean", loss_weight: float = 1.0):
```
If the intention is to follow KL loss API, `reduction` should be `str | None = "batchmean"`

sovrasov reviewed (changes requested):
Thanks @Jyc323 for you contribution, I have a question:
- This PR adds a new loss implementation but does not reveal how OTX users can utilize that loss when solving their tasks. Could you also add an integration into existing h-cls training pipeline and mention that improvement in OTX documentation? I believe loss choice should be configurable via OTX model recipe.

sovrasov commented:
Code quality checks can be launched locally with `cd lib && tox r -vv -e pre-commit`

Jyc323 committed:
fix code review comments

Jyc323 committed:
fix tox errors

Jyc323 committed:
integrate KL loss via recipe YAML; add new H-label model & classifier

Jyc323 commented:
Hi @kprokofi, as suggested, I created a new example recipe yaml file `efficientnet_v2_kl.yaml` for integrating tree path KL divergence loss. To successfully running the recipe file with engine, I added a new model class `TimmModelHLabelClsWithKL` under `lib/src/otx/backend/native/models/classification/hlabel_models/timm_model.py`, and a new classifier class `KLHLabelClassifier` under `lib/src/otx/backend/native/models/classification/classifier/h_label_classifier.py`. Could you please take a look and share any suggestions?

While running `tox r -vv -e pre-commit`, I encountered an error:
> lib/src/otx/backend/native/utils/utils.py:89: error: Module has no attribute "metrics"  [attr-defined]
> Found 1 error in 1 file (checked 363 source files)git

I'm not sure if it's caused by something on my end. 

Also, could you let me know the appropriate location for the documentation? If possible, please provide a template or example I can refer to.

Thanks!

Jyc323 committed:
merge

Jyc323 commented:
Hi @sovrasov, I merged this PR with the current develop branch. Please let me know the next step. Thanks!

Jyc323 committed:
Merge branch 'develop' into hier_loss

sovrasov commented on library/tests/unit/backend/native/models/classification/losses/test_tree_path_kl_divergence.py:
```diff
@@ -0,0 +1,135 @@
+# Copyright (C) 2025 Intel Corporation
```
This file should be moved to `library/tests/...`, since `lib` folder was renamed to `library`

sovrasov commented on library/src/otx/backend/native/models/classification/hlabel_models/timm_model.py:
```diff
@@ -89,3 +89,66 @@ def _create_model(self, head_config: dict | None = None) -> nn.Module:  # type:
         return OTXv1Helper.load_cls_effnet_v2_ckpt(state_dict, "hlabel", add_prefix)
+
+
+class TimmModelHLabelClsWithKL(OTXHlabelClsModel):
```
It looks like a similar classes should be implemented for all the model files in this folder.

I was quickly thinking about how can we avoid that, but my only thought for now is a refactoring of the parent h-cls classes introducing classifier type generics to generate a bunch of h-cls model classes with different h-label classifiers. That's a solution from the top of my head, one can settle on simple duplication for now.

sovrasov commented on library/src/otx/backend/native/models/classification/classifier/h_label_classifier.py:
```diff
@@ -143,3 +144,87 @@ def _forward_explain(self, images: torch.Tensor) -> dict[str, torch.Tensor | lis
         return outputs
+
+
+class KLHLabelClassifier(HLabelClassifier):
```
New classifier stuff should also be covered with UTs

sovrasov commented:
> Hi @kprokofi, as suggested, I created a new example recipe yaml file `efficientnet_v2_kl.yaml` for integrating tree path KL divergence loss. To successfully running the recipe file with engine, I added a new model class `TimmModelHLabelClsWithKL` under `lib/src/otx/backend/native/models/classification/hlabel_models/timm_model.py`, and a new classifier class `KLHLabelClassifier` under `lib/src/otx/backend/native/models/classification/classifier/h_label_classifier.py`. Could you please take a look and share any suggestions?
> 
> While running `tox r -vv -e pre-commit`, I encountered an error:
> 
> > lib/src/otx/backend/native/utils/utils.py:89: error: Module has no attribute "metrics"  [attr-defined]
> > Found 1 error in 1 file (checked 363 source files)git
> 
> I'm not sure if it's caused by something on my end.
> 
> Also, could you let me know the appropriate location for the documentation? If possible, please provide a template or example I can refer to.
> 
> Thanks!

Thanks @Jyc323 for the recent updates. Once new classifier is available for each classification models and extra unit tests are added, we're almost there: only docs are to be added.
Since we have [an advanced tutorial](https://github.com/open-edge-platform/training_extensions/blob/develop/library/docs/source/guide/tutorials/advanced/hier_metric_collection.rst) for h-cls metrics, we can extend it to an advanced h-cls tutorial by introducing examples on how to use the new loss with the new metrics.

Jyc323 committed:
move files since renaming

Jyc323 committed:
refactor to base.py, add unit test

Jyc323 committed:
Merge branch 'develop' into hier_loss

Jyc323 committed:
fix errors from tox

Jyc323 commented:
Hi @sovrasov, thanks for your suggestion, I carefully thought about it. I refactor `OTXHlabelClsModel` by a dynamic wrapping approach to avoid modifying all child implementations or external calls.
In `base.py`, I added a custom `__getattribute__` hook that automatically wraps each child’s `_create_model()` at runtime. The wrapper injects a common post-processing step that executes immediately after `_create_model()` finishes.
The parent class now transparently runs an additional step, `_finalize_model()`, after model creation. `_finalize_model()` replaces the classifier with `KLHLabelClassifier` when `kl_weight > 0`, enabling hybrid training that combines the Tree-Path KL loss with standard cross-entropy.
Since `kl_weight` is a new hyperparameter, I added `**kwargs` parameter to `OTXHlabelClsModel`. This allows the class to accept additional optional arguments for future extensibility. However, I’m not sure if it reduces type safety or clarity.
I also add more unit tests based on the suggestions.
For the changed files, please take a look and let me know if it aligns with your expectations. If it does, I’ll proceed with adding  the related documentation

Jyc323 committed:
Merge branch 'develop' into hier_loss

sovrasov commented on library/src/otx/backend/native/models/classification/hlabel_models/timm_model.py:
```diff
@@ -89,3 +91,66 @@ def _create_model(self, head_config: dict | None = None) -> nn.Module:  # type:
         return OTXv1Helper.load_cls_effnet_v2_ckpt(state_dict, "hlabel", add_prefix)
+
+
+class TimmModelHLabelClsWithKL(OTXHlabelClsModel):
```
Should a specific classifier for timm be deleted since we have a common one?

sovrasov commented on library/src/otx/backend/native/models/classification/hlabel_models/base.py:
```diff
@@ -59,7 +61,9 @@ def __init__(
         torch_compile: bool = False,
+        **kwargs,
     ) -> None:
+        self.kl_weight = kwargs.get("kl_weight", 0.0)
```
That extra hidden argument looks too implicit: for instance `kl_weight` arg is hidden from the signature, therefore no doc and IDE can not give relevant hints.

Since we have `kl_weight` anyway, I propose to make it a normal argument and describe the related logic in the docstring

sovrasov commented:
> Hi @sovrasov, thanks for your suggestion, I carefully thought about it. I refactor `OTXHlabelClsModel` by a dynamic wrapping approach to avoid modifying all child implementations or external calls. In `base.py`, I added a custom `__getattribute__` hook that automatically wraps each child’s `_create_model()` at runtime. The wrapper injects a common post-processing step that executes immediately after `_create_model()` finishes. The parent class now transparently runs an additional step, `_finalize_model()`, after model creation. `_finalize_model()` replaces the classifier with `KLHLabelClassifier` when `kl_weight > 0`, enabling hybrid training that combines the Tree-Path KL loss with standard cross-entropy. Since `kl_weight` is a new hyperparameter, I added `**kwargs` parameter to `OTXHlabelClsModel`. This allows the class to accept additional optional arguments for future extensibility. However, I’m not sure if it reduces type safety or clarity. I also add more unit tests based on the suggestions. For the changed files, please take a look and let me know if it aligns with your expectations. If it does, I’ll proceed with adding the related documentation

Thanks @Jyc323, I've left a couple of comments to the latest changes

Jyc323 committed:
delete unnecessary class, replace **kwargs with kl_weight (add docstring)

Jyc323 committed:
modify the mock of head_idx_to_logits_range

Jyc323 committed:
update list models pattern

Jyc323 committed:
update list models pattern

Jyc323 committed:
Merge branch 'develop' into hier_loss

Jyc323 commented:
Hi @sovrasov , I resolved the comment, and other errors from tox. Please have a review, thanks!

sovrasov commented:
> Hi @sovrasov , I resolved the comment, and other errors from tox. Please have a review, thanks!

@Jyc323 it looks like ruff check is still failing:
```
ruff.....................................................................Failed
- hook id: ruff
- files were modified by this hook
```

Jyc323 committed:
ruff fix

Jyc323 committed:
fix ruff errors

Jyc323 committed:
Merge branch 'develop' into hier_loss

Jyc323 commented:
Hi, @sovrasov, I have fixed the errors, please have a review, thanks

sovrasov commented:
@Jyc323 now unit tests are failing

Jyc323 committed:
Merge branch 'develop' into hier_loss

Jyc323 committed:
Merge branch 'develop' into hier_loss

Jyc323 commented:
Hi @sovrasov, thanks for the review! The unit tests were passing on my side. Could you please recheck when you get a chance?

leoll2 commented on library/src/otx/backend/native/utils/utils.py:
```diff
@@ -86,7 +86,7 @@ def mock_modules_for_chkpt() -> Iterator[None]:
         sys.modules["otx.core.types.label"] = otx.types.label
         sys.modules["otx.core.model"] = otx.backend.native.models  # type: ignore[attr-defined]
-        sys.modules["otx.core.metrics"] = otx.metrics
+        # sys.modules["otx.core.metrics"] = otx.metrics
```
This change is probably causing this test to fail:

<img width="666" height="529" alt="image" src="https://github.com/user-attachments/assets/1c127dbf-9ed6-4e59-a8a3-d045502826c9" />

Jyc323 committed:
Merge branch 'develop' into hier_loss

Jyc323 committed:
fix errors

Jyc323 commented:
Hi @leoll2, thanks for your review. I restored the utils.py file now. Please let me know if there is any other issues.

leoll2 committed:
fix style (ruff)

sovrasov commented:
Thanks a lot @Jyc323! From the code perspective, we are there, so now could you update the advanced h-cls tutorial you have added in the previous PR to avoid these improvements being silently buried inside the library?

Jyc323 committed:
docs: add advanced tutorial for Tree-Path KL Divergence

Jyc323 committed:
Merge branch 'develop' into hier_loss

Jyc323 commented:
Hi @sovrasov , docs have been updated: added the Tree-Path KL tutorial. Please review when convenient!

sovrasov commented:
@Jyc323 thanks for the contribution, impressive work!
